tracer/src/Datadog.Trace/Propagators/SpanContextPropagator.cs

```diff
@@ -175,14 +175,30 @@ public void Inject<TCarrier>(SpanContext context, TCarrier carrier, Action<TCarr
             return Extract(serializedSpanContext, default(ReadOnlyDictionaryGetter));
         }
 
-        public IEnumerable<KeyValuePair<string, string?>> ExtractHeaderTags<T>(T headers, IEnumerable<KeyValuePair<string, string?>> headerToTagMap, string defaultTagPrefix)
-            where T : IHeadersCollection
+        public void ExtractHeaderTags<THeaders>(ISpan span, THeaders headers, IEnumerable<KeyValuePair<string, string?>> headerToTagMap, string defaultTagPrefix)
+            where THeaders : IHeadersCollection
         {
-            return ExtractHeaderTags(headers, headerToTagMap, defaultTagPrefix, string.Empty);
+            var processor = new SpanTagHeaderTagProcessor(span);
+            ExtractHeaderTags(ref processor, headers, headerToTagMap, defaultTagPrefix, string.Empty);
         }
 
-        public IEnumerable<KeyValuePair<string, string?>> ExtractHeaderTags<T>(T headers, IEnumerable<KeyValuePair<string, string?>> headerToTagMap, string defaultTagPrefix, string useragent)
-            where T : IHeadersCollection
+        public void ExtractHeaderTags<THeaders>(ISpan span, THeaders headers, IEnumerable<KeyValuePair<string, string?>> headerToTagMap, string defaultTagPrefix, string useragent)
+            where THeaders : IHeadersCollection
+        {
+            var processor = new SpanTagHeaderTagProcessor(span);
+            ExtractHeaderTags(ref processor, headers, headerToTagMap, defaultTagPrefix, useragent);
+        }
+
+        public void ExtractHeaderTags<THeaders, TProcessor>(ref TProcessor processor, THeaders headers, IEnumerable<KeyValuePair<string, string?>> headerToTagMap, string defaultTagPrefix)
+            where THeaders : IHeadersCollection
+            where TProcessor : struct, IHeaderTagProcessor
```

Comment:
technically it doesn't have to be, but because I changed the code from allocating and returning an IEnumerable<> to this one, it doesn't make sense to allow a class that would have to be allocated.